Third draft for ZEP(Zarr Enhancement Proposal), a community feedback process for Zarr (#16)
zarr-developers/governance: ZEP/instructions/zep0000.md

# Zarr Community Feeback Process - Zarr Enhancement Proposal (ZEP)

--- 
 
## ZEP 0 — Purpose and process

Author: Sanket Verma <svsanketverma5@gmail.com>

Status: Active

Type: Process

Created: 2022-14-03

## What is ZEP?

ZEP stands for Zarr Enhancement Proposal. A ZEP is a design document providing information to the Zarr community, or describing a new feature for Zarr or its

> processes or environment. The ZEP should provide a concise technical specification of the feature and a rationale for the feature.

Author: Added this.

> We intend ZEPs to be the primary mechanisms for proposing major new features, for collecting community input on an issue, and for documenting the design

Author: Added.

decision that has gone into Zarr. The ZEP author is responsible for building consensus within the community and documenting dissenting opinions.

> Because the ZEPs are maintained as text files in a versioned repository, their revision history is the historical record of the feature proposal.

Author: As discussed during the ZSC call on 14/4 we'd be using a separate repository named `zep` for incoming ZEPs.

## ZEP audience

The typical primary audience for ZEPs is the developers working on Zarr and its various implementations, the Zarr steering council as well as the Zarr community.

The broader Zarr community may also choose to use the process to document expected feature additions and to manage complex design coordination problems that
require collaboration across multiple projects.

## ZEP types

> Standard Track ZEP

Describes a new feature or implementation for Zarr

> > Informational ZEP

Reviewer: ```suggestion
### Informational ZEP
```

Describes a ZEP design issue or provides general guidelines or information to the Zarr community but does not propose a new feature. Informational ZEPs do not 
necessarily represent Zarr community consensus or recommendation, so users and implementers are free to ignore informational ZEPS or follow their advice.

> Process ZEP

Describes a new process around Zarr and its implementations, or proposes a change to(or an event in) a process. Process ZEPs are like Standard Track ZEPs but apply to areas other than the Zarr project itself. 

> They may propose an implementation, but not to Zarr’s codebase; they require community consensus; unlike informational ZEPs, they are more than recommendations, and users are typically not free to ignore them. 

Reviewer 2: Right maybe a voting procedure with a threshold would be appropriate

> They may propose an implementation, but not to Zarr’s codebase; they require community consensus; unlike informational ZEPs, they are more than recommendations, and users are typically not free to ignore them. 

Author: Yes, see [this](https://github.com/MSanKeys963/governance/blob/master/ZEP/instructions/zep0000.md#how-does-a-zep-becomes-accepted).

> Examples include procedures, guidelines, changes to the decision-making process, and changes to the tools or environment used in Zarr’s development. Any meta-ZEP is also considered a Process ZEP.

Author: Done.

## ZEP Workflow

> The ZEP process begins with a new idea for Zarr. It is highly recommended that a single ZEP contain a single key proposal or new idea. Small enhancements or 
> patches often don’t need a ZEP and can be injected into the Zarr development workflow with a pull request to the Zarr ZEP repo. The more focused the ZEP, 
> the more successful it tends to be. If in doubt, split your ZEP into several well-focused ones.

Reviewer: ```suggestion
The ZEP process begins with an idea for a significant change to the Zarr specification.
A significant change is anything that would require action from the Zarr implementations.
Minor changes (e.g. wording clarifications, typo fixes, etc., do not need a ZEP).
Changes to Zarr implementations also do not require a ZEP; those should be managed by each implementation community.
```

Author: Resolved.

Each ZEP must have a champion—someone who writes the ZEP using the style and format described below, shepherds the discussions in the appropriate forums, 
and attempts to build community consensus around the idea. The ZEP champion (a.k.a. Author) should first attempt to ascertain whether the idea is suitable for a ZEP. 
Posting to the [Gitter](https://gitter.im/zarr-developers/community) is the best way to go about doing this.

Vetting an idea publicly before going as far as writing a ZEP is meant to save the potential author time. Asking the Zarr community first if an idea is original 
helps prevent too much time being spent on something that is guaranteed to be rejected based on prior discussion. It also helps to make sure the idea applies to 
the entire community and not just the author. Just because an idea sounds good to the author does not mean it will work for most people.

> Once the champion has asked the Zarr community whether an idea has any chance of acceptance, a draft ZEP should be presented to the appropriate venue mentioned 
> above. This gives the author a chance to flesh out the draft ZEP to make properly formatted, of high quality, and to address initial concerns about the proposal.

Reviewer: ```suggestion
Once the champion has asked the Zarr community whether an idea has any chance of acceptance, a draft ZEP 
should be created as PR to the `zarr-specs` repository.
This gives the author a chance to flesh out the draft ZEP to make properly formatted, of high quality, and to address initial concerns about the proposal.
One the author is ready to share the ZEP, it should be marked as "ready for review".
```

After the PR for the ZEP is in place, a post should be made on the Gitter channel containing the sections up to “Backward compatibility”, to limit discussion 
there to usage and impact. Discussion on the pull request will have a broader scope, also including details of implementation.

> Standards Track ZEPs consist of two parts, a design document and a reference implementation. It is generally recommended that at least a prototype implementation 

Reviewer 3: If `Standards Track` means something other than "impacts all implementations", then I could see adding that as an additional category.

Author: >  If `Standards Track` means something other than "impacts all implementations",

`Standards Track ZEP` are something that introduces a new feature to the Zarr. It's up to the author to mention that their ZEP affects a single implementation or all of the implementations. Whatever the case maybe they need to add prototype implementation for what they've mentioned in their ZEP.

> then I could see adding that as an additional category.

Do you mean having `Standard Track ZEP` for every implementation? Like `Standard Track ZEP - Python`, `Standard Track ZEP - Java` etc.?

> Standards Track ZEPs consist of two parts, a design document and a reference implementation. It is generally recommended that at least a prototype implementation 

Author: Is there anything else you think we could add here which might help the author?

> Standards Track ZEPs consist of two parts, a design document and a reference implementation. It is generally recommended that at least a prototype implementation 

Reviewer: ```suggestion
Specification ZEPs consist of two parts: a changes to the spec itself and the narrative ZEP document explaining the need for the change.
It is also highly recommended that the Specification ZEP be accompanied by an implementation PR in at least one of the Zarr implementation libraries,
as ideas that sound good in principle sometimes turn out to be impractical when subjected to the test of implementation. 
```

Author: Done.

> be co-developed with the ZEP, as ideas that sound good in principle sometimes turn out to be impractical when subjected to the test of implementation. 
> Often it makes sense for the prototype implementation to be made available as PR to the Zarr repo (making sure to appropriately mark the PR as a WIP).

Reviewer: ```suggestion
```

Author: Removed Standard Track ZEPs.

> Often it makes sense for the prototype implementation to be made available as PR to the Zarr repo (making sure to appropriately mark the PR as a WIP).

Author: Removed Standard Track ZEPs.

### Submitting a ZEP

> The proposal should be submitted as a draft ZEP via a GitHub pull request to the zarr-developers/zep directory with the name `zep-<n>.rst` where `<n>` is an appropriately assigned four-digit number (e.g., zep-0000.rst). The draft ZEP must use the [ZEP X - Template and Instructions](https://github.com/MSanKeys963/governance/blob/master/ZEP/template/template.md) file.

Reviewer: ```suggestion
The proposal should be submitted as two pull requests to the `zarr-developers/zarr-specs` repo.
The first PR should contain narrative text of the ZEP in the `zep` directory with the name `zep-<n>.rst` where `<n>` is an appropriately assigned four-digit number (e.g., zep-0000.rst). The draft ZEP must use the [ZEP X - Template and Instructions](https://github.com/MSanKeys963/governance/blob/master/ZEP/template/template.md) file.
For Specification ZEPs, there should be a second PR with the actual proposed changes to the spec.
```

Reviewer 3: I assume if it's happening to a single repo, it could be one PR, no?

Reviewer: That depends on whether we care about "publishing" the ZEP to a website before it is accepted (as done with [NEPs](https://numpy.org/neps/). In that case, we would merge the ZEP text PR first as a "Status: Draft" NEP, while the actual PR implementing the change would remain umerged until the ZEP is accepted.

Alternatively, we may decide that the process of publishing the ZEPs as standalone documents is not necessary for Zarr, and instead the PR, including its description, effectively IS the ZEP. I would not be opposed to that. It might be easier to manage.

Author: Resolved this at the ZSC meeting on 14/4. The decision is to have a separate repo `zep` which would render and display the ZEPs on the webpage which means draft ZEPs would also be displayed on the website with the status as `Draft`.

A few points to consider while submitting your ZEP:

- It should sound complete. The ideas must make technical sense.
- The title should accurately describe the content.
- The ZEP’s language (spelling, grammar, sentence structure etc.) and code style should be correct and conformant.

> The [Zarr Steering Council](https://github.com/zarr-developers/governance/blob/master/GOVERNANCE.md#steering-council) will not unreasonably deny publication of a ZEP. Reasons for denying ZEP include duplication of effort, being technically unsound, not providing proper 

Reviewer 3: I'm not really convinced that this need be the ZSC. I can understand that it may simplify bootstrapping at this point, but I'd encourage others' ideas on a way to make this more inclusive.

Author: I think the reason to deny publication of ZEP is something on which ZSC could act swiftly. Deferring this decision to the community will only result in an extended timeframe.

Also, would love to hear the thoughts of others from ZSC.
Cc: @alimanfoo @jakirkham @rabernat @ryan-williams

Reviewer: Why would we ever deny publication of a ZEP?

Author: Added Zarr Implementations Council too.

motivation or addressing backwards compatibility, or not taking care of Zarr [CODE OF CONDUCT](https://github.com/zarr-developers/.github/blob/main/CODE_OF_CONDUCT.md).

### Discussing a ZEP

As soon as the draft ZEP is committed to the ZEP repository, a discussion thread for the ZEP should be created to provide a central place to discuss and review 
its contents. The ZEP author(s) may select the [GitHub discussion feature](https://github.com/zarr-developers/zarr-python/discussions) in the zarr-python repository, 
open a new issue in the Zarr community repository or start the discussion in the Zarr Gitter channel. The discussion regarding the ZEP should follow Zarr’s 

> [CODE OF CONDUCT](https://github.com/zarr-developers/.github/blob/main/CODE_OF_CONDUCT.md) at all times.

Reviewer: I think we should just discuss the ZEP in the ZEP PR.

Author: I've added multiple avenues for discussions [here](https://github.com/zarr-developers/governance/pull/16).

ZEP authors are responsible for collecting community feedback on a ZEP. However, to avoid long-winded and open-ended discussions, strategies such as soliciting 
private or more narrowly-tailored feedback in the early design phase, collaborating with other community members with expertise in the ZEP’s subject matter, 
and picking appropriately-specialised discussion for the ZEP’s topic should be considered. ZEP authors should use their discretion here.

Once the ZEP is committed to the ZEP repository, substantive issues should generally be discussed on the canonical public thread, as opposed to private 
channels or unrelated venues. This ensures everyone can follow and contribute, avoids fragmenting the discussion, and makes sure it is fully considered as part 
of the ZEP review process. Comments, support, concerns and other feedback on this designated thread are a critical part when reviewing the ZEP.

### Review and Resolution

The possible paths of the status of ZEPs are as follows:

![Flowchart](/ZEP/assets/flowchart.png)

All ZEPs should be created with the `Draft` status.

> Eventually, after the discussion, there may be a consensus that the ZEP should be accepted - see the next section for details. At this point, the status becomes 

Reviewer 3: Again, the definition of this consensus is probably one of the trickiest bits.

A ZEP can also be assigned status `Deferred`. The ZEP author or [Zarr Steering Council](https://github.com/zarr-developers/governance/blob/master/GOVERNANCE.md#steering-council) can assign the ZEP this status when no progress is being made on the ZEP.

A ZEP can also be `Rejected`. Perhaps, after all, is said and done it was not a good idea. It is still important to have a record of this fact. The `Withdrawn` 
status is similar—it means that the ZEP author themselves has decided that the ZEP is a bad idea, or has accepted that a competing proposal is a better alternative.

When a ZEP is `Accepted`, `Rejected`, or `Withdrawn`, the ZEP should be updated accordingly. In addition to updating the status field, at the very least the 
`Resolution` header should be added with a link to the relevant link of the discussion.

ZEPs can also be `Superseded` by a different ZEP, rendering the original obsolete. The `Replaced-By` and `Replaces` headers should be added to the original and 
new ZEPs respectively.
Process ZEPs may also have a status of `Active` if they are never meant to be completed, e.g. ZEP 0 (this ZEP).

### How does a ZEP becomes accepted?

> A ZEP is `Accepted` by the consensus of all interested contributors. We need a concrete way to tell whether consensus has been reached. When you think ZEP is ready to accept, create a new discussion in [zarr-python](https://github.com/zarr-developers/zarr-python/discussions) using ‘General’ as the category with a subject like:

Reviewer 3: How are "interested contributors" defined?

Author: The term "interested contributors" here means all who were involved during the discussion of ZEP, users who will be affected by the change and the subset of the broader community who is interested in accepting/rejecting the change.

Reviewer 3: see https://github.com/zarr-developers/governance/pull/17

> `Proposal to accept ZEP <number>: <title>`

In the body of your discussion, you should:

- Link to the latest version of ZEP,
- Briefly describe any major points of contention and how they were resolved,

> - Include a sentence like: “If there are no substantive objections within 7 days from this post, then the ZEP will be accepted;

Reviewer 3: Slightly concerned about the "silence == acceptance" route.

Reviewer 4: I think some sort of quorum rule is useful here. We're looking at (and have been loosely using) [Martha's rules](https://third-bit.com/files/2020/08/marthas/) for scverse, which says at least half the voting members had to hold a vote for acceptance.

Author: > Slightly concerned about the "silence == acceptance" route.

if you think increasing the timeframe would help then we could do that too but I believe having a fixed timeframe for the community to raise objections would help us in preventing open-ended discussion and reaching conclusions.

May I know what concerns you about "silence"?

Author: > We're looking at (and have been loosely using) [Martha's rules](https://third-bit.com/files/2020/08/marthas/) for scverse, which says at least half the voting members had to hold a vote for acceptance.

This is interesting. Thanks for sharing @ivirshup.

Reviewer 3: > May I know what concerns you about "silence"?

It's hard to know if people are not disagreeing simply because they don't have time to respond.

Author: Added a few lines to tackle this [here](https://github.com/MSanKeys963/governance/blob/master/ZEP/instructions/zep0000.md#how-does-a-zep-becomes-accepted).

Note: Currently, Specification ZEPs only deal with the core storage specification of Zarr. The scope of SPEC ZEPs will be extended in the future.

> Informational ZEP

Describes a ZEP design issue or provides general guidelines or information to the Zarr community but does not propose a new feature. Informational ZEPs do not necessarily represent Zarr community consensus or recommendation, so users and implementers are free to ignore informational ZEPS or follow their advice.

> Process ZEP

> Describes a new process around Zarr and its implementations or proposes a change to(or an event in) a process. They may propose an implementation, but not to Zarr’s codebase; they require community consensus; unlike informational ZEPs, they are more than recommendations, and users are typically not free to ignore them. Examples include procedures, guidelines, changes to the decision-making process, and changes to the tools or environment used in Zarr’s specification development. Any meta-ZEP is also considered a Process ZEP.

Reviewer 5: > They may propose an implementation, but not to Zarr’s codebase;

Not sure what that means, could you give an example here @MSanKeys963? The rest of the explanation of the different ZEPs seem very clear to me :+1:

Reviewer 5: I guess my confusion resulted from associating implementations with code-implementations, thanks for the clarification @MSanKeys963

Because the ZEPs are maintained as text files in a versioned repository, their revision history is the historical record of the feature proposal.

## ZEP audience

The typical primary audience for ZEPs is the developers working on Zarr and its various implementations, the Zarr steering council as well as the Zarr community.

The broader Zarr community may also choose to use the process to document expected feature additions and to manage complex design coordination problems that
require collaboration across multiple projects.

## ZEP types

> Specification ZEP

Specification ZEPs deal with the changes related to Zarr spec. The SPEC ZEP could introduce changes in core specification or can add changes as an extension to the core protocol. The changes related to core specification follow a strict and thorough review process and should be adopted by everyone in the Zarr community, whereas changes made as an extension to the core protocol follow a comparatively simple review process and is not strictly necessary to be adopted by everyone in the Zarr community. The specification ZEPs are of two types:

- Core protocol ZEP

> Describes a ZEP which involves changes in the core specification protocol for Zarr. Core protocol ZEPs(commonly known as Core ZEPs) are a part of Spec ZEPs and apply to [Zarr storage specification](https://github.com/zarr-developers/zarr-specs) and its various implementations under the [Zarr Implementations](https://github.com/zarr-developers/zarr_implementations) GitHub repo. The core protocol ZEPs should be adopted by every implementation of Zarr and, in general, the overall Zarr community. Core ZEPs must go through a thorough review process, including involvement, discussion, and voting from the author of the proposed ZEP, Zarr Steering Council, author of various Zarr implementations, and open-source projects/research groups using Zarr and the general Zarr community. The general advice is that everyone should be made aware of the change by the introduction of core ZEP. Core protocol ZEP require community consensus, and users are typically not free to ignore them.

Reviewer 2: ```suggestion
Describes a ZEP which involves changes in the core specification protocol for Zarr. Core protocol ZEPs (commonly known as Core ZEPs) are a part of Spec ZEPs and apply to [Zarr storage specification](https://github.com/zarr-developers/zarr-specs) and its various implementations under the [Zarr Implementations](https://github.com/zarr-developers/zarr_implementations) GitHub repo. The core protocol ZEPs should be adopted by every implementation of Zarr and, in general, the overall Zarr community. Core ZEPs must go through a thorough review process, including involvement, discussion, and voting from the author of the proposed ZEP, Zarr Steering Council, author of various Zarr implementations, and open-source projects/research groups using Zarr and the general Zarr community. The general advice is that everyone should be made aware of the change by the introduction of core ZEP. Core protocol ZEP require community consensus, and users are typically not free to ignore them.
```

Author: Resolved.

> Describes a ZEP which involves changes in the core specification protocol for Zarr. Core protocol ZEPs(commonly known as Core ZEPs) are a part of Spec ZEPs and apply to [Zarr storage specification](https://github.com/zarr-developers/zarr-specs) and its various implementations under the [Zarr Implementations](https://github.com/zarr-developers/zarr_implementations) GitHub repo. The core protocol ZEPs should be adopted by every implementation of Zarr and, in general, the overall Zarr community. Core ZEPs must go through a thorough review process, including involvement, discussion, and voting from the author of the proposed ZEP, Zarr Steering Council, author of various Zarr implementations, and open-source projects/research groups using Zarr and the general Zarr community. The general advice is that everyone should be made aware of the change by the introduction of core ZEP. Core protocol ZEP require community consensus, and users are typically not free to ignore them.

Reviewer 2: What do we mean by "involvement" on this line?

Author: This mostly means getting them involved in the discussion and voting process by means of creating issues or opening discussion threads or using gitter.

- Protocol extension ZEP

> Describes a ZEP which involves changes in the extension of the core specification protocol for Zarr. Protocol extension ZEPs(commonly known as extension ZEPs) are part of Spec ZEPs and apply to Zarr storage extension specification and various implementations under the Zarr Implementations GitHub repo. The extension ZEPs are designed so that they are not necessary to be adopted by every other Zarr implementation or by the whole community. They follow a lenient review process compared to core ZEPs, which includes involvement, discussions, and voting from the author of the proposed ZEP, Zarr Steering Council, and open-source/research groups affected by the change. The extension ZEPs may or may not include involvement from the author of various Zarr implementations and the general Zarr community. The general idea of having an extension ZEP is to add functionality, data types and features in a much more relaxed and quick manner to accommodate emerging and essential use cases. The general advice is that the Zarr community and anyone else who will be affected by the extension ZEPs should be made aware of the change. Extension ZEPs may or may not require community consensus, and it is up to users to ignore them.

Reviewer 2: We should clarify what we mean by users or use a different term. Maybe implementers if that is what is intended?

> Standard Track ZEP

Describes a new feature or implementation for Zarr

> Informational ZEP

Describes a ZEP design issue or provides general guidelines or information to the Zarr community but does not propose a new feature. Informational ZEPs do not 
necessarily represent Zarr community consensus or recommendation, so users and implementers are free to ignore informational ZEPS or follow their advice.

> Process ZEP

> Describes a new process around Zarr and its implementations, or proposes a change to(or an event in) a process. Process ZEPs are like Standard Track ZEPs but apply to areas other than the Zarr project itself. 

Reviewer: ```suggestion
Describes a new process around Zarr and its implementations, or proposes a change to(or an event in) a process. Process ZEPs are like Standard Track ZEPs but apply to areas other than the Zarr Spec itself. 
```

Author: Resolved.

> Describes a new process around Zarr and its implementations, or proposes a change to(or an event in) a process. Process ZEPs are like Standard Track ZEPs but apply to areas other than the Zarr project itself. 

Reviewer 2: ```suggestion
Describes a new process around Zarr and its implementations, or proposes a change to (or an event in) a process. Process ZEPs are like Standard Track ZEPs but apply to areas other than the Zarr project itself.
```

They may propose an implementation, but not to Zarr’s codebase; they require community consensus; unlike informational ZEPs, they are more than recommendations, and users are typically not free to ignore them. 
Examples include procedures, guidelines, changes to the decision-making process, and changes to the tools or environment used in Zarr’s development. Any meta-ZEP is also considered a Process ZEP.

## ZEP Workflow

The ZEP process begins with a new idea for Zarr. It is highly recommended that a single ZEP contain a single key proposal or new idea. Small enhancements or 
patches often don’t need a ZEP and can be injected into the Zarr development workflow with a pull request to the Zarr ZEP repo. The more focused the ZEP, 
the more successful it tends to be. If in doubt, split your ZEP into several well-focused ones.

Each ZEP must have a champion—someone who writes the ZEP using the style and format described below, shepherds the discussions in the appropriate forums, 
and attempts to build community consensus around the idea. The ZEP champion (a.k.a. Author) should first attempt to ascertain whether the idea is suitable for a ZEP. 

> Posting to the [Gitter](https://gitter.im/zarr-developers/community) is the best way to go about doing this.

Reviewer 2: Given Gitter isn't very searchable would suggest an issue on the spec repo or discussion. Something that makes it easier to track the discussion. Ok if we want to share a link on Gitter. Just wanted to avoid making major decisions there

Author: Added a few more areas for discussion.

# Zarr Community Feedback Process - Zarr Enhancement Proposal (ZEP)

--- 
 
## ZEP 0 — Purpose and process

Author: Sanket Verma <svsanketverma5@gmail.com>

Status: Active

Type: Process

Created: 2022-14-03

## What is ZEP?

ZEP stands for Zarr Enhancement Proposal. A ZEP is a design document providing
information to the Zarr community, describing a modification or enhancement of

> the Zarr storage specification, a new feature for its processes or environment.

Reviewer: ```suggestion
the Zarr specification, a new feature for its processes or environment.
```

Let's try to use consistent language around what we call the spec

Reviewer 3: Agreed. I think this stems from a difference between the v2 and v3 usage. V2 uses "storage specification": https://zarr.readthedocs.io/en/stable/spec/v2.html -- I'd vote for updating to the V3 usage.

Reviewer 2: Agreed. Having "storage" here will also be confusing when discussing stores in the spec. So "Zarr specification" seems simpler

The ZEP should provide specific proposed changes to the Zarr specification and
a narrative rationale for the specification changes.

We intend ZEPs to be the primary mechanism for evolving the spec, collecting
community input on major issues and documenting the design decision that has
gone into Zarr. In addition, the ZEP author is responsible for building
consensus within the community and documenting dissenting opinions.

Because the ZEPs are maintained as text files in a versioned repository, their
revision history is the historical record of the feature proposal.

## ZEP audience

The typical primary audience for ZEPs is the developers working on Zarr and its
various implementations, the Zarr steering council as well as the Zarr
community.

The broader Zarr community may also choose to use the process to document
expected feature additions and to manage complex design coordination problems
that require collaboration across multiple projects.

## ZEP types

> Specification ZEP

> Specification ZEPs deal with the changes related to [Zarr Specifications].

Reviewer: Missing link here

Reviewer 3: These link to https://zarr-specs.readthedocs.io/ for me via the references at the bottom.

Author: The links are in the bottom of the document.

The SPEC ZEP (abbreviation of Specification ZEP) introduces changes in the core
specification of Zarr. The changes related to core specification follow a
strict and thorough review process and should be adopted by everyone in the
Zarr community.

- Core protocol ZEP

Describes a ZEP which involves changes in the core specification of Zarr. Core
protocol ZEPs (commonly known as Core ZEPs) are a part of SPEC ZEPs and apply to
[Zarr Specifications] and its various implementations under the
[zarr_implementations] GitHub repo. The core protocol ZEPs should be adopted by
every implementation of Zarr and, in general, the overall Zarr community. Core
ZEPs must go through a thorough review process, including involvement,
discussion, and voting from the author of the proposed ZEP, Zarr Steering
Council, author(s) of various Zarr implementations, and open-source
projects/research groups using Zarr and the general Zarr community. The general
advice is that everyone should be made aware of changes introduced by core

> ZEPs. Core protocol ZEPs require community consensus, and consumers are typically not

Reviewer: Can we clarify what is meant by "consumers"? That word has not been part of our lexicon so far.

Reviewer 3: Fair enough. Sorry it was a late change. It was previously "users" which I don't think fits. "Developers"?

Author: I think 'users' refers to a broader community where it involves contributors, maintainers and users of the project whereas Zarr developers would mean contributors and maintainers only.

How about using 'Zarr developers' and 'users' both?

Reviewer: Ideally Zarr users don't really have to know much about the spec. Maybe we need the notion of "downstream users" or "dependents of" zarr? Or maybe we just stick with "developers".

Author: > Ideally Zarr users don't really have to know much about the spec.

I just wanted the ZEP process to be inclusive and welcoming for everyone and not leave out any part(s) of the Zarr community while designing/discussing a ZEP.

I've added this at the start:

> WHERE:
> 
> - Developers refer to contributors and maintainers of the project
> - User(s) refers to an individual or group of individuals or the broader
> community using the project in any way.

and used the term **developers** and **users** instead of *consumers*.

Let me know what you think of this. ^

Reviewer 2: What about Zarr implementers?

Author: I think implementers come under the term 'developer'. Developer is more like an umbrella terminology here.

free to ignore them.

> Note: Currently, Specification ZEPs only deal with the core storage

Reviewer: ```suggestion
Note: Currently, Specification ZEPs only deal with the core
```

Is the the "Core protocol", the "Core storage specification" or the "Core specification". Let's pick one and stick with it.

Author: I'd like to go with "Core storage specification".

Reviewer 3: We will need to cross check whatever is set here with https://zarr-specs.readthedocs.io/. There, "storage" is more about things like file v. zip v. sharding. But I'm definitely for getting this language correct consistently across the involved repos.

Author: Let's go with a more general term for now? 'Core specification'?

Reviewer: That's fine. But how is the "core specification" distinct from just the "specification"? Where in the specification does it say "this is the core part"?

Reviewer 2: Agreed. Let's just go with "specification". Seems like that is what is proposed above as well ( https://github.com/zarr-developers/governance/pull/16#discussion_r861841411 )

specification of Zarr. The scope of SPEC ZEPs will be extended in the future.

> Informational ZEP

Describes a ZEP design issue or provides general guidelines or information to
the Zarr community but does not propose a new feature. Informational ZEPs do
not necessarily represent Zarr community consensus or recommendation, so users
and implementers are free to ignore informational ZEPS or follow their advice.

> Process ZEP

Describes a new process around Zarr and its implementations or proposes a
change to (or an event in) a process. They may propose an implementation, but

> not to Zarr’s codebase; they require community consensus; unlike informational

Reviewer: There is no "Zarr codebase". There is only the spec and the implementations. This needs to be revised.

Reviewer 3: :+1:

Author: Done. 💯

> ZEPs, they are more than recommendations, and users are typically not free to

Reviewer: What do Zarr users care about our process? Surely we are talking about developers here, no?

```suggestion
ZEPs, they are more than recommendations, and Zarr developers are typically not free to
```

ignore them. Examples include procedures, guidelines, changes to the
decision-making process, and changes to the tools or environment used in Zarr’s
specification development. Any meta-ZEP is also considered a Process ZEP.

## ZEP Workflow

The ZEP process begins with a new idea for Zarr. It is highly recommended that
a single ZEP contain a single key proposal or new idea. Small enhancements or
patches often don’t need a ZEP and can be injected into the Zarr development
workflow with a pull request to the Zarr ZEP repo. The more focused the ZEP,
the more successful it tends to be. If in doubt, split your ZEP into several
well-focused ones.

Each ZEP must have a champion -- someone who writes the ZEP using the style and
format described below, shepherds the discussions in the appropriate forums,
and attempts to build community consensus around the idea. The ZEP champion
(a.k.a. Author) should first attempt to ascertain whether the idea is suitable

> for a ZEP. Posting to [Gitter] or creating an issue on the [community] repository

Reviewer: Let's clarify that ZEPs can have multiple co-authors.

Author: Added a separate line for this. ^
Done. 💯

deferring some features to later ZEPs), as this status can lead to version
compatibility challenges in the wider Zarr ecosystem.

A ZEP can also be assigned status `Deferred`. The ZEP author or [Zarr Steering
Council] can assign the ZEP this status when no progress is being made on the
ZEP.

A ZEP can also be `Rejected`. Perhaps, after all, is said and done it was not a
good idea. It is still important to have a record of this fact. The `Withdrawn`
status is similar—it means that the ZEP author themselves has decided that the
ZEP is a bad idea, or has accepted that a competing proposal is a better
alternative.

When a ZEP is `Accepted`, `Rejected`, or `Withdrawn`, the ZEP should be updated
accordingly. In addition to updating the status field, at the very least the
`Resolution` header should be added with a link to the relevant link of the
discussion.

ZEPs can also be `Superseded` by a different ZEP, rendering the original
obsolete. The `Replaced-By` and `Replaces` headers should be added to the
original and new ZEPs respectively. Process ZEPs may also have a status of
`Active` if they are never meant to be completed, e.g. ZEP 0 (this ZEP).

> ### How does a ZEP becomes accepted?

Reviewer: ```suggestion
### How does a ZEP become accepted?
```

for Specifications ZEPs using ‘General’ and in zarr-developers/zep for other
ZEPs and as the category with a subject like:

> `Proposal to accept ZEP <number>: <title>`

In the body of your discussion, you should:

- Link to the latest version of ZEP,
- Briefly describe any major points of contention and how they were resolved,
- Include a sentence like: “If there are no substantive objections within 7 days
  from this post, then the ZEP will be accepted;

After you create the discussion, you should make sure to link the newly created
thread in the `Discussion` section of the ZEP, so that people can find it later.

Generally, the ZEP author will be the one to create this post, but anyone can
do it – the important thing is to make sure that everyone knows when a ZEP is
on the verge of acceptance, and give them a final chance to respond. If
there’s some special reason to extend this final comment period beyond
`7 days`, then that’s fine, just say so in the post. You shouldn’t do less
than `7 days`, because sometimes people are travelling or similar and need
some time to respond.

There may be a case that a ZEP didn’t attract needed attention towards it, the

> engagement from the community is low, and `seven days` pass by. In this case,

Reviewer: ```suggestion
engagement from the community is low, and `7 days` pass by. In this case,
```

Let's use consistent spelling

to your follow-up discussion thread.

If there are substantive objections, then the ZEP remains in `Draft` state,
discussion continues as normal, and it can be proposed for acceptance again
later once the objections are resolved.

In unusual cases, the [Zarr Steering Council] may be asked to decide whether a
controversial ZEP is `Accepted`.

### Maintenance

In general, SPEC ZEPs are no longer modified after they have reached the Final
state. The changes made to the Zarr’s core specification in the
[zarr-specs] repository are considered the ultimate reference. However,
finalised SPEC ZEPS may be updated with minor changes as needed.

Process ZEPs may be updated over time to reflect changes to development
practices and other details. The precise process followed in these cases will
depend on the nature and purpose of the ZEP being updated.

## ZEP Format

ZEPs are UTF-8 encoded text files using the
[reStructureText](http://docutils.sourceforge.net/rst.html) format. Please see
the [ZEP X - Template and Instructions][template.md] file and the
[reStructuredTextPrimer](https://www.sphinx-doc.org/en/stable/rest.html) for

> more information.

Reviewer 4: Very minor comment, and the rest of this looks great.

This says ZEPs are in rst, but this document and the template are in markdown. These should probably be aligned. As a note, this is repeated multiple times throughout the document.

Would be nice if they aligned on markdown 😉

Author: Hi @ivirshup.

ZEP0 documents(files in this PR) will be converted into .rst files in a separate repo named `zep`. The markdown files are here so that everyone can read and review them easily.

I'd be taking care of this ^ after this PR is finalised and there are no more comments/objections.

Reviewer 4: Makes sense.

Wondering if it might then also make sense to keep a copy of the markdown template around so people writing ZEPs can do the same?

Author: Yeah, that sounds right. I think we can leave the ZEP 0 files (in .md format) here too.

### How does a ZEP becomes accepted?

> A ZEP is `Accepted` by the consensus of all interested contributors. We need a concrete way to tell whether the agreement has been reached. 

Reviewer 3: For those following along, https://github.com/zarr-developers/governance/pull/17 is relevant to the discussion of "interested contributors"
